frontend/src/metabase/visualizer/utils/get-initial-state-for-card-data-source.ts

```diff
@@ -43,31 +32,135 @@ export function getInitialStateForCardDataSource(
       }
 
       if (Array.isArray(originalValue)) {
-        return [
-          setting,
-          originalValue.map(originalColumnName => {
-            const index = columns.findIndex(
-              col => col.name === originalColumnName,
+        const mappedColumns = originalValue
+          .map(originalColumnName => {
+            const columnInfo = columnInfos.find(
+              info => info.columnRef.originalName === originalColumnName,
             );
-            return state.columns[index].name;
-          }),
-        ];
-      } else {
-        const index = columns.findIndex(col => col.name === originalValue);
-        if (!state.columns[index]) {
-          return;
-        }
+            return columnInfo?.columnRef.name;
+          })
+          .filter(isNotNull);
 
-        return [setting, state.columns[index]?.name];
+        return mappedColumns.length > 0 ? [setting, mappedColumns] : null;
+      } else {
+        const columnInfo = columnInfos.find(
+          info => info.columnRef.originalName === originalValue,
+        );
+        return columnInfo?.columnRef.name
+          ? [setting, columnInfo.columnRef.name]
+          : null;
       }
     })
     .filter(isNotNull);
 
+  return Object.fromEntries(entries);
+}
+
+function processColumnsForDataSource(
+  dataSource: VisualizerDataSource,
+  columns: DatasetColumn[],
+  state: VisualizerHistoryItem,
+): ColumnInfo[] {
+  const columnInfos: ColumnInfo[] = [];
+
+  columns.forEach(column => {
+    const columnRef = createVisualizerColumnReference(
+      dataSource,
+      column,
+      extractReferencedColumns(state.columnValuesMapping),
+    );
+
+    const processedColumn = copyColumn(
+      columnRef.name,
+      column,
+      dataSource.name,
+      state.columns,
+    );
+
+    state.columns.push(processedColumn);
+    state.columnValuesMapping[columnRef.name] = [columnRef];
+
+    columnInfos.push({
+      columnRef,
+      column: processedColumn,
+    });
+  });
+
+  return columnInfos;
+}
+
+export function getInitialStateForCardDataSource(
+  card: Card,
+  columns: DatasetColumn[],
+): VisualizerHistoryItem {
+  const state: VisualizerHistoryItem = {
+    display: card.display,
+    columns: [],
+    columnValuesMapping: {},
+    settings: {},
+  };
+
+  const dataSource = createDataSource("card", card.id, card.name);
+  const columnInfos = processColumnsForDataSource(dataSource, columns, state);
+  const mappedSettings = mapColumnVizSettings(card, columnInfos);
+
   state.settings = {
     ...card.visualization_settings,
-    ...Object.fromEntries(entries),
+    ...mappedSettings,
     "card.title": card.name,
   };
 
   return state;
 }
+
+export function getInitialVisualizerStateForMultipleSeries(
```

Comment:
This should probably live in its own file, and be named so it's eaier to understand the link with `getInitialStateForCardDataSource`. Maybe something like:
 - `getInitialStateForCardDataSource` -> `getInitialStateForSingleSeriesCard`
 - `getInitialStateForMultipleSeries` -> `getInitialStateForMultipleSeriesCard`
if the two functions live at the same level of course. Right now it's a little hard to tell wether they're two functions that respond to two different cases, or if one is higher up in terms of responsibility than the other.

Reply:
I moved it into a separate file https://github.com/metabase/metabase/pull/56043/files#diff-06143777ed9b0cad3c445c2ccbe50d681614193cf3e0853b5656c1417318d65b
Regarding naming, I think card suffix can be a bit misleading in common terms of the codebase because each series itself contains a card and the "card" suffix in `getInitialStateForMultipleSeriesCard` would've looked like there is only one card. Not sure if I should change `getInitialStateForCardDataSource` as it is not created from a series object. Wdyt?